src/databasetasks.h

```diff
@@ -47,6 +47,7 @@ class DatabaseTasks : public ThreadHolder<DatabaseTasks>
 		void threadMain();
 	private:
 		void runTask(const DatabaseTask& task);
+		void flushInternal();
```

Comment:
I think you need to get rid of the locking in shutdown because as it is right now, it will hang on shutdown.